bpf/lib/policy_log.h

```diff
@@ -0,0 +1,82 @@
+/*
+ *  Copyright (C) 2016-2019 Authors of Cilium
+ *
+ *  This program is free software; you can redistribute it and/or modify
+ *  it under the terms of the GNU General Public License as published by
+ *  the Free Software Foundation; either version 2 of the License, or
+ *  (at your option) any later version.
+ *
+ *  This program is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with this program; if not, write to the Free Software
+ *  Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
+ */
+/*
+ * Policy verdict notification via perf event ring buffer.
+ *
+ * API:
+ * void send_policy_verdict_notify4(skb, remote_label, dir, verdict, match_type)
+ *
+ * If POLICY_VERDICT_NOTIFY is not defined, the API will not be compiled in.
+ */
+
+#ifndef __LIB_POLICY_LOG__
+#define __LIB_POLICY_LOG__
+
+#include "common.h"
+
+#ifdef POLICY_VERDICT_NOTIFY
+struct policy_verdict_notify {
+	NOTIFY_CAPTURE_HDR
+	__u32	remote_label;
+	__u8	action;
+	__u8	match_type;
```

Comment:
Sure, I can make that change. I would like to get some suggestions on if more information should be included in this structure. I think it depends on what fields are useful for the quick rule association in the user plane. match_type was suggested by @tgraf

Reply:
I could imagine a path where userspace `cilium-agent` creates identifiers for each policymap entry and pushes these down into the policymap, then when doing policy verdict tracing with these messages, those identifiers are passed back up in the monitor message.

If these are allocated by the agent, then 16 bits (65K ids) seems like plenty of uniqueness for an individual policy rule on a single node.
If these end up being centrally managed, then I could imagine desire for this to be a UUID.

Reply:
I considered this way of implementation, but it is pretty complicated because a traffic can match multiple policies and it incurs extra effort to manage the identifier in this case. I feel it is probably not the direction Cilium wants to pursue. On the other hand, how difficult it is to add fields later? or we leave some padding there for future use ?

Reply:
I would imagine the policy multiplexing would occur in userspace to avoid creating a list of identifiers in the dataplane. But I agree, this is complicated.

In the absence of other feedback regarding upcoming features, I would be inclined to just round it to the nearest allocation boundary and leave it at this.

Reply:
Updated the patch after the offline discussion. Added proto and port fields to the event and rearranged padding.